proposed/0007-Address_relocation.md

```diff
@@ -0,0 +1,40 @@
+- Type enhancement
+- Related components routing, safe_vault
+- Start Date: 21-09-2015
+- RFC PR: (leave this empty)
+- Issue number: (leave this empty)
+
+# Summary
+
+Prevent a potential attack where a node gains overall/majority influence in a group setting.
+
+# Motivation
+
+The possibility 'may' exist for a node on the network to generate off-line derived identities close to a specific target and to then connect nodes whose resultant network relocated identities actually lie close to that target. In such cases the owner of that node will dominate, or at least have a dominant influence within, the group close to that target. This RFC proposes an amendment to the current process to prevent this possibility arising in practice.
+
+# Detailed design
+
+A node joining the SAFE network generates a new cryptographic key pair to access the network for each session. The generated public key, K, is then cryptographically hashed using some hash function H to give H(K). The node then sends a message, containing K, to the group H(K) requesting a network identifier for the session. On receipt of such a request each node belonging to the group hashes the concatenation of K with the two closest nodes, N1, N2 say, in xor distance from K present in their routing tables, giving H(K + N1 + N2) = N, the identifier the node will use on the network for the session.
+
+Having computed N, each node in the group H(K) sends a message, containing K, to the close group of N to confirm address relocation has been applied for the node owning K. A sufficient condition for the receiving group to resolve the message is for a quorum in the sending group to agree on the two closest nodes. Each node in the group H(K) further sends a message, containing N, to the joining node confirming the network address it now occupies.
+
+In it's current form the process ensures that a joining node cannot occupy a position of it's own choosing. However, in a semi-stable network state it may be possible given the joining node knows the two closest nodes, N3, N4 say, in it’s joined group to generate keys that are relocated close to some identifier, I say, and thus obtain group influence at that location. It is the purpose of this RFC to avoid that possibility by storing the two close node identities, N3 and N4, at each node belonging to the joined group in order to reject relocations involving either of those nodes if they occur more than once for the participating nodes in a given session.
+
+## Implementation
+
+The `RoutingNode` function `handle_request_network_name` calculates a relocated name for a node joining the network. The utils.rs function `pub fn calculate_relocated_name(mut close_nodes: Vec<::NameType>, original_name: &::NameType) -> Result<::NameType, ::error::RoutingError>`, takes the close nodes from the routing table and original name, K, as arguments, returning H(K + N1 + N2) = N, in the notation above, and is used for this purpose.
```

Comment:
> One possible solution is to make pmid_node a separate vault instance using pre-determined network address derived from account, meanwhile other personas kept in another vault instance using dynamic-determined network address.

I think this is the best approach.  That Vault doesn't even need to be the same process, or even the same GitHub project.  In some regards the PMID persona is more like a client anyway: it's a single entity on the network which can only validate its actions by signing - it's not a member of a group of nodes.  So if it has no authority to mutate data, there's no need to ever apply address relocation to it.

In fact, running it as a separate process would probably be useful for low-power, low-spec, or bandwidth-limited scenarios, since such a device could possibly run the personas Vault quite comfortably (and hence be "giving back" to the network) and not run the storage Vault.

Another possible solution is that relocation is permanent - i.e. the network-generated value `N` forms part of the PMID packet for the new node.

This means the relocation only happens once when a new Vault joins for the first time.

The downside with this is the longer the Vault exists, the less "validatable" `N` will become - the nodes which created `N` will become lower and lower in the list of closest to the original address as the network grows.